File: packages/native/src/hash.ts
/**
 * Native xxHash64 module with JS fallback
 *
 * Provides high-performance hashing for Merkle tree operations.
 * Falls back to JS implementation if native module is unavailable.
 */

/**
 * Native hash module interface
 */
export interface NativeHashModule {
  xxhash64(data: Buffer | Uint8Array, seed?: bigint | number): bigint;
  xxhash64AsNumber(data: Buffer | Uint8Array, seed?: number): number;
  xxhash64Batch(
    buffers: (Buffer | Uint8Array)[],
    seed?: bigint | number
  ): bigint[];
  xxhash64BatchAsNumbers(
    buffers: (Buffer | Uint8Array)[],
    seed?: number
  ): number[];
  XxHash64State: new (seed?: bigint | number) => XxHash64StateInstance;
}

export interface XxHash64StateInstance {
  update(data: Buffer | Uint8Array): this;
  digest(): bigint;
  digestAsNumber(): number;
  reset(): this;
}

let nativeModule: NativeHashModule | null = null;
let loadAttempted = false;
let loadError: Error | null = null;

/**
 * Try to load native hash module.
 * Returns null if native module is not available.
 */
export function getNativeHash(): NativeHashModule | null {
  if (loadAttempted) return nativeModule;

  loadAttempted = true;

  try {
    // Try release build first
    nativeModule = require('../build/Release/topgun_hash.node');
    return nativeModule;
  } catch (e1) {
    try {
      // Try debug build
      nativeModule = require('../build/Debug/topgun_hash.node');
      return nativeModule;
    } catch (e2) {
      loadError = e1 as Error;
      // Silent fallback - don't spam console
      return null;
    }
  }
}

/**
 * Check if native module is available.
 */
export function isNativeHashAvailable(): boolean {
  return getNativeHash() !== null;
}

/**
 * Get the load error if native module failed to load.
 */
export function getNativeHashLoadError(): Error | null {
  getNativeHash(); // Ensure load was attempted
  return loadError;
}

// ============ JS Fallback Implementation ============

/**
 * FNV-1a hash (current implementation in core).
 * Used as fallback - produces 32-bit hash.
 */
function fnv1aHash(str: string): number {
  let hash = 0x811c9dc5;
  for (let i = 0; i < str.length; i++) {
    hash ^= str.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

/**
 * JS implementation of xxHash64.
 * This is a simplified version that provides similar distribution.
 * Note: Not bit-compatible with C xxHash64, use only as fallback.
 */
function xxhash64JS(data: Uint8Array, seed: bigint = 0n): bigint {
  const PRIME64_1 = 0x9e3779b185ebca87n;
  const PRIME64_2 = 0xc2b2ae3d27d4eb4fn;
  const PRIME64_3 = 0x165667b19e3779f9n;
  const PRIME64_4 = 0x85ebca77c2b2ae63n;
  const PRIME64_5 = 0x27d4eb2f165667c5n;

  const len = BigInt(data.length);
  let h64: bigint;

  if (data.length >= 32) {
    // Process in 32-byte chunks
    let v1 = seed + PRIME64_1 + PRIME64_2;
    let v2 = seed + PRIME64_2;
    let v3 = seed;
    let v4 = seed - PRIME64_1;

    let i = 0;
    while (i + 32 <= data.length) {
      v1 = xxh64Round(v1, readU64(data, i));
      v2 = xxh64Round(v2, readU64(data, i + 8));
      v3 = xxh64Round(v3, readU64(data, i + 16));
      v4 = xxh64Round(v4, readU64(data, i + 24));
      i += 32;
    }

    h64 = rotl64(v1, 1n) + rotl64(v2, 7n) + rotl64(v3, 12n) + rotl64(v4, 18n);
    h64 = mergeRound(h64, v1);
    h64 = mergeRound(h64, v2);
    h64 = mergeRound(h64, v3);
    h64 = mergeRound(h64, v4);
  } else {
    h64 = seed + PRIME64_5;
  }

  h64 += len;

  // Process remaining bytes
  let remaining = data.length % 32;
  let offset = data.length - remaining;

  while (remaining >= 8) {
    const k1 = xxh64Round(0n, readU64(data, offset));
    h64 ^= k1;
    h64 = rotl64(h64, 27n) * PRIME64_1 + PRIME64_4;
    offset += 8;
    remaining -= 8;
  }

  while (remaining >= 4) {
    h64 ^= BigInt(readU32(data, offset)) * PRIME64_1;
    h64 = rotl64(h64, 23n) * PRIME64_2 + PRIME64_3;
    offset += 4;
    remaining -= 4;
  }

  while (remaining > 0) {
    h64 ^= BigInt(data[offset]) * PRIME64_5;
    h64 = rotl64(h64, 11n) * PRIME64_1;
    offset++;
    remaining--;
  }

  // Final mix
  h64 ^= h64 >> 33n;
  h64 *= PRIME64_2;
  h64 ^= h64 >> 29n;
  h64 *= PRIME64_3;
  h64 ^= h64 >> 32n;

  return h64 & 0xffffffffffffffffn;
}

function readU64(data: Uint8Array, offset: number): bigint {
  return (
    BigInt(data[offset]) |
    (BigInt(data[offset + 1]) << 8n) |
    (BigInt(data[offset + 2]) << 16n) |
    (BigInt(data[offset + 3]) << 24n) |
    (BigInt(data[offset + 4]) << 32n) |
    (BigInt(data[offset + 5]) << 40n) |
    (BigInt(data[offset + 6]) << 48n) |
    (BigInt(data[offset + 7]) << 56n)
  );
}

function readU32(data: Uint8Array, offset: number): number {
  return (
    data[offset] |
    (data[offset + 1] << 8) |
    (data[offset + 2] << 16) |
    (data[offset + 3] << 24)
  );
}

function rotl64(x: bigint, r: bigint): bigint {
  return ((x << r) | (x >> (64n - r))) & 0xffffffffffffffffn;
}

function xxh64Round(acc: bigint, input: bigint): bigint {
  const PRIME64_1 = 0x9e3779b185ebca87n;
  const PRIME64_2 = 0xc2b2ae3d27d4eb4fn;
  acc += input * PRIME64_2;
  acc &= 0xffffffffffffffffn;
  acc = rotl64(acc, 31n);
  acc *= PRIME64_1;
  return acc & 0xffffffffffffffffn;
}

function mergeRound(acc: bigint, val: bigint): bigint {
  const PRIME64_1 = 0x9e3779b185ebca87n;
  const PRIME64_4 = 0x85ebca77c2b2ae63n;
  val = xxh64Round(0n, val);
  acc ^= val;
  acc = acc * PRIME64_1 + PRIME64_4;
  return acc & 0xffffffffffffffffn;
}

/**
 * JS fallback for xxhash64.
 */
export function xxhash64Fallback(
  data: Buffer | Uint8Array,
  seed: bigint = 0n
): bigint {
  const uint8 = data instanceof Buffer ? new Uint8Array(data) : data;
  return xxhash64JS(uint8, seed);
}

/**
 * JS fallback returning 32-bit number.
 */
export function xxhash64FallbackAsNumber(
  data: Buffer | Uint8Array,
  seed: number = 0
): number {
  const hash = xxhash64Fallback(data, BigInt(seed));
  return Number(hash & 0xffffffffn);
}

/**
 * JS fallback for batch hash.
 */
export function xxhash64BatchFallback(
  buffers: (Buffer | Uint8Array)[],
  seed: bigint = 0n
): bigint[] {
  return buffers.map((buf) => xxhash64Fallback(buf, seed));
}

/**
 * JS fallback for batch hash returning numbers.
 */
export function xxhash64BatchFallbackAsNumbers(
  buffers: (Buffer | Uint8Array)[],
  seed: number = 0
): number[] {
  return buffers.map((buf) => xxhash64FallbackAsNumber(buf, seed));
}

// ============ Unified API ============

/**
 * Compute xxHash64 of data.
 * Uses native module if available, otherwise falls back to JS.
 */
export function xxhash64(
  data: Buffer | Uint8Array,
  seed: bigint | number = 0
): bigint {
  const native = getNativeHash();
  if (native) {
    return native.xxhash64(data, typeof seed === 'number' ? BigInt(seed) : seed);
  }
  return xxhash64Fallback(data, typeof seed === 'number' ? BigInt(seed) : seed);
}

/**
 * Compute xxHash64 and return as 32-bit number.
 * This is more efficient when BigInt precision is not needed.
 */
export function xxhash64AsNumber(
  data: Buffer | Uint8Array,
  seed: number = 0
): number {
  const native = getNativeHash();
  if (native) {
    return native.xxhash64AsNumber(data, seed);
  }
  return xxhash64FallbackAsNumber(data, seed);
}

/**
 * Compute xxHash64 for multiple buffers.
 */
export function xxhash64Batch(
  buffers: (Buffer | Uint8Array)[],
  seed: bigint | number = 0
): bigint[] {
  const native = getNativeHash();
  if (native) {
    return native.xxhash64Batch(
      buffers,
      typeof seed === 'number' ? BigInt(seed) : seed
    );
  }
  return xxhash64BatchFallback(
    buffers,
    typeof seed === 'number' ? BigInt(seed) : seed
  );
}

/**
 * Compute xxHash64 for multiple buffers, returning 32-bit numbers.
 */
export function xxhash64BatchAsNumbers(
  buffers: (Buffer | Uint8Array)[],
  seed: number = 0
): number[] {
  const native = getNativeHash();
  if (native) {
    return native.xxhash64BatchAsNumbers(buffers, seed);
  }
  return xxhash64BatchFallbackAsNumbers(buffers, seed);
}

/**
 * Create streaming hash state.
 */
export function createXxHash64State(
  seed: bigint | number = 0
): XxHash64StateInstance {
  const native = getNativeHash();
  if (native) {
    return new native.XxHash64State(seed);
  }
  return new JsXxHash64State(typeof seed === 'number' ? BigInt(seed) : seed);
}

/**
 * JS fallback streaming implementation
 */
class JsXxHash64State implements XxHash64StateInstance {
  private chunks: Uint8Array[] = [];
  private seed: bigint;

  constructor(seed: bigint) {
    this.seed = seed;
  }

  update(data: Buffer | Uint8Array): this {
    this.chunks.push(new Uint8Array(data));
    return this;
  }

  digest(): bigint {
    const totalLength = this.chunks.reduce((sum, chunk) => sum + chunk.length, 0);
    const combined = new Uint8Array(totalLength);
    let offset = 0;
    for (const chunk of this.chunks) {
      combined.set(chunk, offset);
      offset += chunk.length;
    }
    return xxhash64Fallback(combined, this.seed);
  }

  digestAsNumber(): number {
    return Number(this.digest() & 0xffffffffn);
  }

  reset(): this {
    this.chunks = [];
    return this;
  }
}

// ============ String Hashing (for compatibility with current code) ============

/**
 * Hash a string using xxHash64.
 * Returns 32-bit number for compatibility with existing hashString() usage.
 */
export function hashString(str: string): number {
  const native = getNativeHash();
  if (native) {
    return native.xxhash64AsNumber(Buffer.from(str));
  }
  // Use FNV-1a as fallback for string hashing (faster for short strings)
  return fnv1aHash(str);
}

/**
 * Hash a string using xxHash64, returning full 64-bit BigInt.
 */
export function hashStringBigInt(str: string, seed: bigint = 0n): bigint {
  const native = getNativeHash();
  if (native) {
    return native.xxhash64(Buffer.from(str), seed);
  }
  return xxhash64Fallback(Buffer.from(str), seed);
}

File: packages/native/src/hash.cc
/**
 * Native xxHash64 implementation for TopGun
 *
 * Provides high-performance hashing for Merkle tree operations.
 * xxHash64 is 10-20x faster than SHA-256 for small inputs.
 */

#include <napi.h>

// xxHash header-only implementation
#define XXH_INLINE_ALL
#include "xxhash.h"

/**
 * Compute xxHash64 of a buffer.
 * @param buffer - Input data (Buffer or Uint8Array)
 * @param seed - Optional seed (default: 0)
 * @returns BigInt with 64-bit hash
 */
Napi::Value XxHash64(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();

    // Validate arguments
    if (info.Length() < 1) {
        Napi::TypeError::New(env, "Expected at least 1 argument")
            .ThrowAsJavaScriptException();
        return env.Null();
    }

    if (!info[0].IsBuffer() && !info[0].IsTypedArray()) {
        Napi::TypeError::New(env, "Expected Buffer or TypedArray")
            .ThrowAsJavaScriptException();
        return env.Null();
    }

    // Get buffer data
    uint8_t* data;
    size_t length;

    if (info[0].IsBuffer()) {
        auto buffer = info[0].As<Napi::Buffer<uint8_t>>();
        data = buffer.Data();
        length = buffer.Length();
    } else {
        auto typedArray = info[0].As<Napi::TypedArray>();
        data = static_cast<uint8_t*>(typedArray.ArrayBuffer().Data()) +
               typedArray.ByteOffset();
        length = typedArray.ByteLength();
    }

    // Get optional seed
    uint64_t seed = 0;
    if (info.Length() > 1 && info[1].IsBigInt()) {
        bool lossless;
        seed = info[1].As<Napi::BigInt>().Uint64Value(&lossless);
    } else if (info.Length() > 1 && info[1].IsNumber()) {
        seed = static_cast<uint64_t>(info[1].As<Napi::Number>().Int64Value());
    }

    // Compute hash
    XXH64_hash_t hash = XXH64(data, length, seed);

    // Return as BigInt for full 64-bit precision
    return Napi::BigInt::New(env, hash);
}

/**
 * Compute xxHash64 and return as 32-bit number (truncated).
 * Useful when BigInt overhead is not needed.
 */
Napi::Value XxHash64AsNumber(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();

    if (info.Length() < 1 || (!info[0].IsBuffer() && !info[0].IsTypedArray())) {
        Napi::TypeError::New(env, "Expected Buffer or TypedArray")
            .ThrowAsJavaScriptException();
        return env.Null();
    }

    uint8_t* data;
    size_t length;

    if (info[0].IsBuffer()) {
        auto buffer = info[0].As<Napi::Buffer<uint8_t>>();
        data = buffer.Data();
        length = buffer.Length();
    } else {
        auto typedArray = info[0].As<Napi::TypedArray>();
        data = static_cast<uint8_t*>(typedArray.ArrayBuffer().Data()) +
               typedArray.ByteOffset();
        length = typedArray.ByteLength();
    }

    uint64_t seed = 0;
    if (info.Length() > 1 && info[1].IsNumber()) {
        seed = static_cast<uint64_t>(info[1].As<Napi::Number>().Int64Value());
    }

    XXH64_hash_t hash = XXH64(data, length, seed);

    // Truncate to 32 bits for compatibility with current FNV-1a usage
    return Napi::Number::New(env, static_cast<uint32_t>(hash & 0xFFFFFFFF));
}

/**
 * Compute xxHash64 for multiple buffers in batch.
 * More efficient than calling XxHash64 repeatedly due to reduced JS<->C++ overhead.
 * @param buffers - Array of Buffers
 * @param seed - Optional seed
 * @returns Array of BigInt hashes
 */
Napi::Value XxHash64Batch(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();

    if (!info[0].IsArray()) {
        Napi::TypeError::New(env, "Expected Array of Buffers")
            .ThrowAsJavaScriptException();
        return env.Null();
    }

    auto buffers = info[0].As<Napi::Array>();
    uint32_t count = buffers.Length();

    uint64_t seed = 0;
    if (info.Length() > 1 && info[1].IsBigInt()) {
        bool lossless;
        seed = info[1].As<Napi::BigInt>().Uint64Value(&lossless);
    } else if (info.Length() > 1 && info[1].IsNumber()) {
        seed = static_cast<uint64_t>(info[1].As<Napi::Number>().Int64Value());
    }

    Napi::Array results = Napi::Array::New(env, count);

    for (uint32_t i = 0; i < count; i++) {
        Napi::Value item = buffers.Get(i);

        if (!item.IsBuffer() && !item.IsTypedArray()) {
            results.Set(i, Napi::BigInt::New(env, static_cast<uint64_t>(0)));
            continue;
        }

        uint8_t* data;
        size_t length;

        if (item.IsBuffer()) {
            auto buffer = item.As<Napi::Buffer<uint8_t>>();
            data = buffer.Data();
            length = buffer.Length();
        } else {
            auto typedArray = item.As<Napi::TypedArray>();
            data = static_cast<uint8_t*>(typedArray.ArrayBuffer().Data()) +
                   typedArray.ByteOffset();
            length = typedArray.ByteLength();
        }

        XXH64_hash_t hash = XXH64(data, length, seed);
        results.Set(i, Napi::BigInt::New(env, hash));
    }

    return results;
}

/**
 * Batch hash returning 32-bit numbers.
 */
Napi::Value XxHash64BatchAsNumbers(const Napi::CallbackInfo& info) {
    Napi::Env env = info.Env();

    if (!info[0].IsArray()) {
        Napi::TypeError::New(env, "Expected Array of Buffers")
            .ThrowAsJavaScriptException();
        return env.Null();
    }

    auto buffers = info[0].As<Napi::Array>();
    uint32_t count = buffers.Length();

    uint64_t seed = 0;
    if (info.Length() > 1 && info[1].IsNumber()) {
        seed = static_cast<uint64_t>(info[1].As<Napi::Number>().Int64Value());
    }

    Napi::Array results = Napi::Array::New(env, count);

    for (uint32_t i = 0; i < count; i++) {
        Napi::Value item = buffers.Get(i);

        if (!item.IsBuffer() && !item.IsTypedArray()) {
            results.Set(i, Napi::Number::New(env, 0));
            continue;
        }

        uint8_t* data;
        size_t length;

        if (item.IsBuffer()) {
            auto buffer = item.As<Napi::Buffer<uint8_t>>();
            data = buffer.Data();
            length = buffer.Length();
        } else {
            auto typedArray = item.As<Napi::TypedArray>();
            data = static_cast<uint8_t*>(typedArray.ArrayBuffer().Data()) +
                   typedArray.ByteOffset();
            length = typedArray.ByteLength();
        }

        XXH64_hash_t hash = XXH64(data, length, seed);
        results.Set(i, Napi::Number::New(env, static_cast<uint32_t>(hash & 0xFFFFFFFF)));
    }

    return results;
}

/**
 * Create streaming hash state.
 * For hashing data incrementally.
 */
class XxHash64State : public Napi::ObjectWrap<XxHash64State> {
public:
    static Napi::Object Init(Napi::Env env, Napi::Object exports) {
        Napi::Function func = DefineClass(env, "XxHash64State", {
            InstanceMethod("update", &XxHash64State::Update),
            InstanceMethod("digest", &XxHash64State::Digest),
            InstanceMethod("digestAsNumber", &XxHash64State::DigestAsNumber),
            InstanceMethod("reset", &XxHash64State::Reset),
        });

        exports.Set("XxHash64State", func);
        return exports;
    }

    XxHash64State(const Napi::CallbackInfo& info)
        : Napi::ObjectWrap<XxHash64State>(info) {
        seed_ = 0;
        if (info.Length() > 0 && info[0].IsBigInt()) {
            bool lossless;
            seed_ = info[0].As<Napi::BigInt>().Uint64Value(&lossless);
        } else if (info.Length() > 0 && info[0].IsNumber()) {
            seed_ = static_cast<uint64_t>(info[0].As<Napi::Number>().Int64Value());
        }
        state_ = XXH64_createState();
        XXH64_reset(state_, seed_);
    }

    ~XxHash64State() {
        if (state_) {
            XXH64_freeState(state_);
        }
    }

    Napi::Value Update(const Napi::CallbackInfo& info) {
        Napi::Env env = info.Env();

        if (!info[0].IsBuffer() && !info[0].IsTypedArray()) {
            Napi::TypeError::New(env, "Expected Buffer or TypedArray")
                .ThrowAsJavaScriptException();
            return env.Undefined();
        }

        uint8_t* data;
        size_t length;

        if (info[0].IsBuffer()) {
            auto buffer = info[0].As<Napi::Buffer<uint8_t>>();
            data = buffer.Data();
            length = buffer.Length();
        } else {
            auto typedArray = info[0].As<Napi::TypedArray>();
            data = static_cast<uint8_t*>(typedArray.ArrayBuffer().Data()) +
                   typedArray.ByteOffset();
            length = typedArray.ByteLength();
        }

        XXH64_update(state_, data, length);
        return info.This();
    }

    Napi::Value Digest(const Napi::CallbackInfo& info) {
        Napi::Env env = info.Env();
        XXH64_hash_t hash = XXH64_digest(state_);
        return Napi::BigInt::New(env, hash);
    }

    Napi::Value DigestAsNumber(const Napi::CallbackInfo& info) {
        Napi::Env env = info.Env();
        XXH64_hash_t hash = XXH64_digest(state_);
        return Napi::Number::New(env, static_cast<uint32_t>(hash & 0xFFFFFFFF));
    }

    Napi::Value Reset(const Napi::CallbackInfo& info) {
        XXH64_reset(state_, seed_);
        return info.This();
    }

private:
    XXH64_state_t* state_;
    uint64_t seed_;
};

// Module initialization
Napi::Object Init(Napi::Env env, Napi::Object exports) {
    exports.Set("xxhash64", Napi::Function::New(env, XxHash64));
    exports.Set("xxhash64AsNumber", Napi::Function::New(env, XxHash64AsNumber));
    exports.Set("xxhash64Batch", Napi::Function::New(env, XxHash64Batch));
    exports.Set("xxhash64BatchAsNumbers", Napi::Function::New(env, XxHash64BatchAsNumbers));
    XxHash64State::Init(env, exports);
    return exports;
}

NODE_API_MODULE(topgun_hash, Init)
